include/deal.II/numerics/nonlinear.h

```diff
@@ -0,0 +1,582 @@
+ *               Vector<double> &      residual) {...}
+ *
+ * nonlinear_solver.setup_jacobian =
+ *           [&](const Vector<double> &current_u,
+ *               const Vector<double> &current_f) {...}
+ *
+ * nonlinear_solver.solve_with_jacobian =
+ *           [&](const Vector<double> &rhs,
+ *               Vector<double> &      dst,
+ *               const double tolerance) {...}
+ *
+ * // Calling the @p solve function with an initial guess.
+ * nonlinear_solver.solve(current_solution);
+ * @endcode
+ */
+template <typename VectorType = Vector<double>>
+class NonlinearSolverSelector
+{
+public:
+  /**
+   * Additional data that will be sent through the NonlinearSolverSelector
+   * class, then into the specific nonlinear solver classes themselves.
+   * Many of these parameters can also be foud in the documentation for
+   * @p KINSOL and @p NOX.
+   */
+  class AdditionalData
+  {
+  public:
+    /**
+     * NonlinearSolverSelector solution strategy. The solver types included in
+     * this class both use a Newton-Krylov solver with a line search and even
+     * Picard iterations (KINSOL).
+     */
+    enum SolutionStrategy
+    {
+      /**
+       * Standard Newton iteration.
+       */
+      newton,
+      /**
+       * Newton iteration with linesearch.
+       */
+      linesearch,
+      /**
+       * Picard iteration.
+       */
+      picard,
+    };
+
+    enum SolverType
+    {
+      /**
+       * Default parameter, will use whatever solver is available
+       * with KINSOL as the priority.
+       */
+      automatic,
+      /**
+       * KINSOL nonlinear solver, part of the SUNDIALS package.
+       */
+      kinsol,
+      /**
+       * NOX nonlinear solver, part of the TRILINOS package.
+       */
+      nox,
+    };
+
+    /**
+     * Initialization parameters for NonlinearSolverSelector.
+     *
+     * @param solver_type Nonlinear solver type, can be 'auto', 'kinsol', or 'nox'.
+     * @param maximum_non_linear_iterations Maximum number of nonlinear
+     * iterations. This parameter is shared between KINSOL and NOX.
+     * @param function_tolerance %Function norm stopping tolerance.
+     */
+    AdditionalData(const SolverType & solver_type                   = automatic,
+                   const unsigned int maximum_non_linear_iterations = 200,
+                   const double       function_tolerance            = 0.0);
+
+    /**
+     * The type of nonlinear solver to use. The default value is set to 'auto',
+     * which will use either KINSOL or NOX depending on which package is
+     * installed, KINSOL having priority.
+     */
+    SolverType solver_type;
+
+    /**
+     * The solution strategy to use. For this class, you can choose from
+     * SolutionStrategy::newton, SolutionStrategy::linesearch, or
+     * SolutionStrategy::picard. More details on this can be found on the
+     * @p KINSOL documentation.
+     */
+    SolutionStrategy strategy = linesearch;
```

Comment:
You've got a funny combination of member variables here, some of which can be initialized via the constructor, and some are default-initialized here and need to be set by hand if one doesn't want the default. In all other places, we typically list all members in the constructor argument list.

Reply:
So do you suggest we put more into the constructor argument? I simply did what @tjhei suggested by setting defaults to parameters more specific to either solver.

Reply:
The point I wanted to make is this:
* If you have lots of parameters that only pertain to one solver or another, then what's the point of the whole class? It only makes sense setting these parameters if you know that you want to use one solver or another, but then you could as well save yourself the trouble and just use that solver. I recognize that others might feel differently about this because the approach you use here still enables you to write code that at least at a high level looks like it's independent of what solver one actually uses.
* My preference would therefore be to identify the subset of parameters that are common to all solvers, and only allow setting those.
* Then have the constructor argument list reflect the *complete* list of member variables in the class, so that they can *all* be set via the constructor. This is perhaps also an odd design decision, but it is what we do for all other `AdditionalData` classes. The constructor arguments would then have default values that match the initializers you provide here to member variables.